picard/cluster.py

```diff
@@ -245,23 +245,83 @@ def cluster(files, threshold):
             artist_max = 0
             artist_id = None
             artist_hist = {}
+            main_artist = ""
+            do_all = False
+            do_cluster = True
+            to_remove = []
             for track_id in album:
+                artist = tracks[track_id][0]
+                if main_artist is "":
+                    main_artist = artist
+
                 cluster = artist_cluster_engine.getClusterFromId(
                     tracks[track_id][0])
-                if cluster is not None:
+
+                # if it isn't the first track the user hasn't chosen an action to do for all
+                if artist is not main_artist:
+                    if not do_all:
+                        do_cluster, do_all = Cluster.cluster_warning(files, track_id, album)
+                    if not do_cluster:
+                        to_remove.append(track_id)
+
+                if do_cluster and cluster is not None:
                     cnt = artist_hist.get(cluster, 0) + 1
                     if cnt > artist_max:
                         artist_max = cnt
                         artist_id = cluster
                     artist_hist[cluster] = cnt
 
+            for id in to_remove:
+                album.remove(id)
+
             if artist_id is None:
                 artist_name = u"Various Artists"
             else:
                 artist_name = artist_cluster_engine.getClusterTitle(artist_id)
 
             yield album_name, artist_name, (files[i] for i in album)
 
+    @staticmethod
+    def cluster_warning(files, track_id, album):
+        QMessageBox = QtGui.QMessageBox
+        QCheckBox = QtGui.QCheckBox
+        QRadioButton = QtGui.QRadioButton
+
+        title = _(u"Album Artist Conflict")
+        text = _(u"This track shares an album title with a cluster, "
+            "but does not share an artist name. Would you still like to "
+            "cluster this track?\n\n")
+
+        # Conflicting Track
+        artist_name = files[track_id].metadata["artist"]
+        album_name = files[track_id].metadata["album"]
+        song_title = files[track_id].metadata["title"]
+        text = text + album_name + '\n' + song_title + '\n' + artist_name
+
+
+        msg = QMessageBox(QMessageBox.Question, title, text)
+        layout = msg.layout()
+        no = msg.addButton("No", QMessageBox.NoRole)
+        yes = msg.addButton("Yes", QMessageBox.YesRole)
+        msg.setDefaultButton(yes)
+
+        cluster_list = "Current cluster:"
+        for cluster_id in album:
+            if cluster_id < track_id:
+                artist = files[cluster_id].metadata["artist"]
+                title = files[cluster_id].metadata["title"]
+                track = files[cluster_id].metadata["tracknumber"]
+                cluster_list = cluster_list + '\n' + '{:4}{:12}{:10}'.format(track, title, artist)
```

Comment:
Instead of concatenating strings this way, use a list, append elements to it, and join them after the loop.